bindings/rust/s2n-tls-hyper/src/stream.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use hyper::rt::{Read, ReadBufCursor, Write};
+use hyper_util::{
+    client::legacy::connect::{Connected, Connection},
+    rt::TokioIo,
+};
+use s2n_tls::connection::Builder;
+use s2n_tls_tokio::TlsStream;
+use std::{
+    io::Error,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+pub enum MaybeHttpsStream<T, B>
+where
+    T: Read + Write + Connection + Unpin,
+    B: Builder,
+    <B as Builder>::Output: Unpin,
+{
+    Https(TokioIo<TlsStream<TokioIo<T>, B::Output>>),
+}
+
+impl<T, B> Connection for MaybeHttpsStream<T, B>
+where
```

Comment:
I don't trust myself without my IDE in front of my, but are these bounds necessary? It seems like they should be inferred given that they are already on the enum definitions.

Reply:
I was able to remove some of the bounds, like the `Connection` bound wasn't needed when implementing `Read` and `Write`. But it didn't like me removing the `Builder` bounds. Though I'm not sure I entirely understand why:

```
error[E0277]: the trait bound `B: s2n_tls::pool::Pool` is not satisfied
  --> s2n-tls-hyper/src/stream.rs:43:27
   |
43 | impl<T, B> Connection for MaybeHttpsStream<T, B>
   |                           ^^^^^^^^^^^^^^^^^^^^^^ the trait `s2n_tls::pool::Pool` is not implemented for `B`
   |
   = note: required because of the requirements on the impl of `s2n_tls::connection::Builder` for `B`
```

Reply:
Unfortunately I think this falls under the "literally will not compile" case. The `TlsStream` struct has bounds on its types:
https://github.com/aws/s2n-tls/blob/d538d0a7031f01a692c7f0dd76fbe08110b509cd/bindings/rust/s2n-tls-tokio/src/lib.rs#L142-L145

`TokioIo` only implements `AsyncRead` and `AsyncWrite` when `T` implements hyper's `Read` and `Write`:
https://github.com/hyperium/hyper-util/blob/53aadac50d03fb93f0cb585e67c1c738afb7e2eb/src/rt/tokio.rs#L147-L151

So I can't provide `TokioIo<T>` to `TlsStream` unless `T` implements hyper's `Read` and `Write`:
```
error[E0277]: the trait bound `T: hyper::rt::Read` is not satisfied
   --> s2n-tls-hyper/src/stream.rs:39:19
    |
39  |     Https(TokioIo<TlsStream<TokioIo<T>, B::Output>>),
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `hyper::rt::Read` is not implemented for `T`
    |
    = note: required because of the requirements on the impl of `tokio::io::async_read::AsyncRead` for `TokioIo<T>`
note: required by a bound in `TlsStream`
```

Reply:
I was able to remove the bounds on the `HttpsConnector` struct though!

Reply:
Hmm, is this perhaps an indication that we should remove the bounds from the TlsStream struct?

Reply:
Ah, I didn't consider that...

Reply:
I had some trouble removing the bounds in the TlsStream struct as well. I'll look into this for a future PR, since it probably wouldn't make sense to modify the TlsStream struct in this PR anyway.